docs: twister: Add section explaining how to choose platform scope (#75962)
zephyrproject-rtos/zephyr: doc/develop/test/twister.rst

  testing relating keywords to provide best coverage for the features of this
  board.

> .. _twister_defautl_testing_board:

Reviewer: ```suggestion
.. _twister_default_testing_board:
```

Most everyday users will run with no arguments.

Selecting platform scope:
*************************

One of the key features of Twister is its ability to decide on which platforms a given
test scenario should run. This behavior has its roots in Twister being developed as
a test runner for Zephyr's CI. With hundreds of available platforms and thousands of
tests, the testing tools should be able to adapt the scope and select/filter out what
is relevant and what is not.

Twister is always preparing an initial list of platforms in scope for a given test,
based on command line arguments and the test's configuration file. Then, platforms
that don't fulfill the required conditions are filtered out from the scope.

> Using ``--force-platform`` turns off filtering.

Reviewer 2: This is not clear to me. What are "the required conditions" exactly? What does turning the filtering off mean?

If a platform is in my `platform_allow` list but not in the default platform list, will it be run if I use `--force-platform`?

Also, the twister CLI help say the following:

```
Force testing on selected platforms, even if they are excluded in the test configuration (testcase.yaml).
```

It doesn't sound the same as what you describe to me. From what I understand it's just a way to ignore `platform_exclude`.

Reviewer 3: `--force-platform` allows you to run test on a platform that is not listed in a `platform_allow` list. This flag is rather used to tests / debug platforms, but finally tested platform should be added to `platform_allow` list, if you want to have it in regression.

Author: I added more clarification about "required conditions" and "filtering" and "--force-platform".

> If a platform is in my platform_allow list but not in the default platform list, will it be run if I use --force-platform?
It depends on what other arguments you've used.  

I believe all logic required to answer this is presented in the PR. I tried to explain multiple times (it is also already part of the docs at https://docs.zephyrproject.org/latest/develop/test/twister.html#tests at `platform_allow`) that platform_allow shouldn't be treated as a scope defining parameter but a filter. Therefore, I am against describing this specific combinations since IMO it is legitimizing misuse of platform_allow.

Command line arguments define the initial scope in the following way:

* ``-p/--platform <platform_name>`` (can be used multiple times): only platforms
  passed with this argument;
* ``--all``: all available platforms;
* ``-G/--integration``: all platforms from an `integration_platforms` list in
  a given test configuration file. If a test has no `integration_platforms`
  the "scope presumption"* will happen;
* No scope argument: `"scope presumption"` will happen.

`"Scope presumption"`: A list of Twister's default platforms is used as
the initial list. If nothing is left after the filtration, a platform_allow list

> is used as the initial scope.

Reviewer: Looks good, but could you expand some more on this section to explain the behavior as @theob-pro was trying to do in https://github.com/zephyrproject-rtos/zephyr/pull/75699?

Reviewer 2: Yes please, you describe the command line argument but could you also describe the keys in the test case file and the "default" testing mechanism of boards.

Author: > Yes please, you describe the command line argument but could you also describe the keys in the test case file and the "default" testing mechanism of boards.

Those are already part of the documentation. I added links to relevant sections.

> could you expand some more on this section to explain the behavior as @theob-pro was trying to do in https://github.com/zephyrproject-rtos/zephyr/pull/75699?

Sorry, but no. As in other comment: I believe all logic required to answer this is presented in the PR. I tried to explain multiple times (it is also already part of the docs at https://docs.zephyrproject.org/latest/develop/test/twister.html#tests at `platform_allow`) that platform_allow shouldn't be treated as a scope defining parameter but a filter. Therefore, I am against describing this specific combinations since IMO it is legitimizing misuse of platform_allow.

Most everyday users will run with no arguments.

Selecting platform scope
************************

One of the key features of Twister is its ability to decide on which platforms a given
test scenario should run. This behavior has its roots in Twister being developed as
a test runner for Zephyr's CI. With hundreds of available platforms and thousands of
tests, the testing tools should be able to adapt the scope and select/filter out what
is relevant and what is not.

> Twister is always preparing an initial list of platforms in scope for a given test,

Reviewer: ```suggestion
Twister always prepares an initial list of platforms in scope for a given test,
```

> based on command line arguments and the :ref:`test configuration file <test>`. Then,

Reviewer: What does this reference?

platforms that don't fulfill the conditions required in the configuration yaml
(e.g. minimum ram) are filtered out from the scope.

> Using ``--force-platform`` allows to override filtering based on ``platform_allow``,

Reviewer: Is this what you mean here?
```suggestion
Using ``--force-platform`` allows to override filtering and instead base it on ``platform_allow``,
```

``platform_exclude``, ``arch_allow`` and ``arch_exclude`` keys in test configuration
files.

Command line arguments define the initial scope in the following way:

* ``-p/--platform <platform_name>`` (can be used multiple times): only platforms
  passed with this argument;
* ``-l/--all``: all available platforms;
* ``-G/--integration``: all platforms from an `integration_platforms` list in
  a given test configuration file. If a test has no `integration_platforms`

>   the "scope presumption"* will happen;

Reviewer: ```suggestion
  `"scope presumption"` will happen;
```

* No scope argument: `"scope presumption"` will happen.

`"Scope presumption"`: A list of Twister's :ref:`default platforms <twister_defautl_testing_board>`

> is used as the initial list. If nothing is left after the filtration, a platform_allow list

Reviewer: ```suggestion
is used as the initial list. If nothing is left after the filtration, the ``platform_allow`` list
```

Most everyday users will run with no arguments.

Selecting platform scope
************************

One of the key features of Twister is its ability to decide on which platforms a given
test scenario should run. This behavior has its roots in Twister being developed as
a test runner for Zephyr's CI. With hundreds of available platforms and thousands of
tests, the testing tools should be able to adapt the scope and select/filter out what
is relevant and what is not.

Twister always prepares an initial list of platforms in scope for a given test,

> based on command line arguments and the :ref:`test configuration file <test_config_args>`. Then,

Reviewer: As you are pointing directly to the configuration items, I would change this to
```suggestion
based on command line arguments and the :ref:`test's configuration <test_config_args>`. Then,
```
